Cython/Utility/ObjectHandling.c

```diff
@@ -79,12 +87,17 @@ static int __Pyx_IternextUnpackEndCheck(PyObject *retval, Py_ssize_t expected) {
 
 /////////////// UnpackTuple2.proto ///////////////
 
+#if CYTHON_ASSUME_SAFE_MACROS
 #define __Pyx_unpack_tuple2(tuple, value1, value2, is_tuple, has_known_size, decref_tuple) \
     (likely(is_tuple || PyTuple_Check(tuple)) ? \
         (likely(has_known_size || PyTuple_GET_SIZE(tuple) == 2) ? \
             __Pyx_unpack_tuple2_exact(tuple, value1, value2, decref_tuple) : \
             (__Pyx_UnpackTupleError(tuple, 2), -1)) : \
         __Pyx_unpack_tuple2_generic(tuple, value1, value2, has_known_size, decref_tuple))
+#else
+static CYTHON_INLINE int __Pyx_unpack_tuple2(
+    PyObject* tuple, PyObject** value1, PyObject** value2, int is_tuple, int has_known_size, int decref_tuple);
+#endif
```

Comment:
It seems worth generally unfolding this lengthy macro into the inline function you wrote, and just special casing the size check.